modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java

```diff
@@ -326,29 +292,82 @@ ThreadPool getThreadPool() {
         return threadPool;
     }
 
-    static class Stats {
+    enum Operation {
+        GET_OBJECT("GetObject"),
+        LIST_OBJECTS("ListObjects"),
+        PUT_OBJECT("PutObject"),
+        PUT_MULTIPART_OBJECT("PutMultipartObject"),
+        DELETE_OBJECTS("DeleteObjects"),
+        ABORT_MULTIPART_OBJECT("AbortMultipartObject");
 
-        final AtomicLong listCount = new AtomicLong();
+        private final String value;
 
-        final AtomicLong getCount = new AtomicLong();
+        Operation(String value) {
+            this.value = value;
+        }
+    }
 
-        final AtomicLong putCount = new AtomicLong();
+    class StatsCollectors {
+        final ConcurrentMap<String, IgnoreNoResponseMetricsCollector> collectors;
```

Comment:
No need for a concurrent map here, we populate it up front. An `EnumMap` would be best.

Reply:
The stats are keyed by a pair of enums that is `Operation/Purpose`, e.g. `GetObject/ClusterState`. The two sets of Enums track different things. For example, the `BlobContainer.compareAndExchangeRegister()` method can be invoked for `Snapshot` purpose while the method internally uses both `ListObjects` and `PutMultipartObject` operations. It results into two entries in stats collection, i.e. `ListObjects/Snapshot` and `PutMultipartObject/Snapshot`. So it is not possible to have it as a EnumMap. 

We populate this map upfront mainly for BWC because existing API always reports all operations regardless whether the operation is actually used or not. Also for BWC reasons, we skip `Snapshot` in the stats key so that `ListObjects/Snapshot` is just `ListObjects`. The code initialise stats collections for purposes other than Snapshot on the fly. This also means they, e.g. `ListObjects/ClusterState` won't be initialised and not visible at all for Stateful.

I hope this makes sense.

Reply:
Ah sorry I see now, completely missed this the first time through.